code/modules/language/language_holder.dm

```diff
@@ -1,141 +1,275 @@
+				spoken_languages -= language
+		. = TRUE
 
-/datum/language_holder/proc/remove_all_languages()
-	languages.Cut()
+/// Removes every language and optionally sets omnitongue false. If a non default source is supplied, only removes that source.
+/datum/language_holder/proc/remove_all_languages(source = LANGUAGE_ALL, omni = FALSE)
+	for(var/language in GLOB.all_languages)
+		remove_language(language, TRUE, TRUE, source)
+	if(omni)
+		omnitongue = FALSE
+	return TRUE
 
-/datum/language_holder/proc/has_language(datum/language/dt)
-	if(is_type_in_typecache(dt, languages))
-		return LANGUAGE_KNOWN
-	else
-		var/atom/movable/AM = get_atom()
-		var/datum/language_holder/L = AM.get_language_holder(shadow=FALSE)
-		if(L != src)
-			if(is_type_in_typecache(dt, L.shadow_languages))
-				return LANGUAGE_SHADOWED
-	return FALSE
+/// Adds a single language or list of languages to the blocked language list.
+/datum/language_holder/proc/add_blocked_language(languages, source = LANGUAGE_MIND)
+	if(!islist(languages))
+		languages = list(languages)
+	for(var/language in languages)
+		if(!blocked_languages[language])
+			blocked_languages += language
+			blocked_languages[language] = list()
+		blocked_languages[language] |= source
+	return TRUE
+
+/// Removes a single language or list of languages from the blocked language list.
+/datum/language_holder/proc/remove_blocked_language(languages, source = LANGUAGE_MIND)
+	if(!islist(languages))
+		languages = list(languages)
+	for(var/language in languages)
+		if(blocked_languages[language])
+			if(source == LANGUAGE_ALL)
+				blocked_languages -= language
+			else
+				blocked_languages[language] -= source
+				if(!blocked_languages[language].len)
+					blocked_languages -= language
+	return TRUE
 
-/datum/language_holder/proc/copy_known_languages_from(thing, replace=FALSE)
-	var/datum/language_holder/other
-	if(istype(thing, /datum/language_holder))
-		other = thing
-	else if(ismovableatom(thing))
-		var/atom/movable/AM = thing
-		other = AM.get_language_holder()
-	else if(istype(thing, /datum/mind))
-		var/datum/mind/M = thing
-		other = M.get_language_holder()
+/// Checks if you have the language. If spoken is true, only checks if you can speak the language.
+/datum/language_holder/proc/has_language(language, spoken = FALSE)
+	if(language in blocked_languages)
+		return FALSE
+	if(spoken)
+		return language in spoken_languages
+	return language in understood_languages
+
+/// Checks if you can speak the language. Tongue limitations should be supplied as an argument.
+/datum/language_holder/proc/can_speak_language(language, tongue = TRUE)
+	if((omnitongue || tongue) && has_language(language, TRUE))
+		return TRUE
+	return FALSE
 
-	if(replace)
-		src.remove_all_languages()
+/// Returns selected language if it can be spoken, or decides, sets and returns a new selected language if possible.
+/datum/language_holder/proc/get_selected_language()
+	if(selected_language && can_speak_language(selected_language))
```

Comment:
Shouldn't this be taking tongue restrictions into account ?

Reply:
Yes, forgot that later in the proc too. Good catch

Reply:
tongue check moved to can_speak_language

Reply:
tongue check moved to can_speak_language